### PR TITLE
#3098 initial build of presentation server

### PR DESCRIPTION
**Pull request by planetf1**
Signed-off-by: Nigel Jones <nigel.l.jones+git@gmail.com>

See #3098 which has a more complete 'todo' list for changes in this area

This is an initial build for a new maven module  **org.odpi.egeria:presentation-server**
npm: 6.14.5 (current LTS)
node: 12.16.0 (12.18.0 is LTS, but breaks with babel)

Node is downloaded and run only from within nodejs/node - it does NOT use system version to maintain consistency

Currently a jar is produced with 3 files:
www/favicon.ico
www/index.html
www/bundle.js

as produced by webpack

The plan for this PR is to perform one kind of build, and to create a deployable artifact (which I will test) - another commit will use the files generated above and include in an assembly.

Expect more refactoring in later PRs.

### CODE REVIEW

**Comment by planetf1** [comment]
Question for reviewers - is it worth merging just with this very basic 'build' capability?
Is the output what you'd expect?

```
jonesn:presentation-server/ (issue3098*) $ mvn clean install                                                                       [17:48:14]
[INFO] Scanning for projects...
[INFO]
[INFO] ----------------< org.odpi.egeria:presentation-server >-----------------
[INFO] Building Egeria Presentation Server 2.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:3.1.0:clean (default-clean) @ presentation-server ---
[INFO] Deleting /Users/jonesn/IdeaProjects/egeria/open-metadata-implementation/user-interfaces/presentation-server/target
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:display-info (display-info) @ presentation-server ---
[INFO] Maven Version: 3.6.3
[INFO] JDK Version: 14.0.1 normalized as: 14.0.1
[INFO] OS Info: Arch: x86_64 Family: mac Name: mac os x Version: 10.15.5
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-versions) @ presentation-server ---
[INFO]
[INFO] --- apache-rat-plugin:0.13:check (rat-check) @ presentation-server ---
[INFO] RAT will not execute since it is configured to be skipped via system property 'rat.skip'.
[INFO]
[INFO] --- frontend-maven-plugin:1.10.0:install-node-and-npm (install node and npm) @ presentation-server ---
[INFO] Node v12.16.0 is already installed.
[INFO] NPM 6.14.5 is already installed.
[INFO]
[INFO] --- frontend-maven-plugin:1.10.0:npm (npm-install) @ presentation-server ---
[INFO] Running 'npm install --prefix .' in /Users/jonesn/IdeaProjects/egeria/open-metadata-implementation/user-interfaces/presentation-server/nodejs
[INFO] npm WARN presentation-server@1.0.0 No repository field.
[INFO]
[INFO] added 2 packages, removed 141 packages, updated 84 packages and audited 1319 packages in 8.055s
[INFO]
[INFO] 36 packages are looking for funding
[INFO]   run `npm fund` for details
[INFO]
[INFO] found 46 vulnerabilities (44 low, 1 moderate, 1 high)
[INFO]   run `npm audit fix` to fix them, or `npm audit` for details
[INFO]
[INFO] --- frontend-maven-plugin:1.10.0:webpack (webpack-build) @ presentation-server ---
[INFO] Running 'webpack.js --mode development' in /Users/jonesn/IdeaProjects/egeria/open-metadata-implementation/user-interfaces/presentation-server/nodejs
[INFO] clean-webpack-plugin: /Users/jonesn/IdeaProjects/egeria/open-metadata-implementation/user-interfaces/presentation-server/nodejs/dist has been removed.
[INFO] Hash: 295a07f9e756968e7bc7
[INFO] Version: webpack 4.43.0
[INFO] Time: 52514ms
[INFO] Built at: 06/04/2020 17:52:15
[INFO]       Asset       Size  Chunks             Chunk Names
[INFO]   bundle.js   23.1 MiB    main  [emitted]  main
[INFO] favicon.ico   43.6 KiB          [emitted]
[INFO]  index.html  496 bytes          [emitted]
[INFO] Entrypoint main = bundle.js
[INFO] [0] multi babel-polyfill ./src/client/index.js 40 bytes {main} [built]
[INFO] [./node_modules/css-loader/dist/cjs.js!./node_modules/sass-loader/dist/cjs.js!./src/client/app.scss] 397 bytes {main} [built]
[INFO] [./node_modules/css-loader/dist/cjs.js!./node_modules/sass-loader/dist/cjs.js!./src/client/index.scss] 456 KiB {main} [built]
[INFO] [./node_modules/webpack/buildin/global.js] (webpack)/buildin/global.js 472 bytes {main} [built]
[INFO] [./src/client/App.js] 744 bytes {main} [built]
[INFO] [./src/client/Frame.js] 4.31 KiB {main} [built]
[INFO] [./src/client/app.scss] 1.19 KiB {main} [built]
[INFO] [./src/client/auth/Login.css] 1.08 KiB {main} [built]
[INFO] [./src/client/auth/login.js] 4 KiB {main} [built]
[INFO] [./src/client/contexts/IdentificationContext.js] 2.88 KiB {main} [built]
[INFO] [./src/client/images/Egeria_glossary_author_32.js] 350 bytes {main} [built]
[INFO] [./src/client/images/Egeria_logo_color.js] 389 bytes {main} [built]
[INFO] [./src/client/images/Egeria_logo_white_110.js] 338 bytes {main} [built]
[INFO] [./src/client/index.js] 290 bytes {main} [built]
[INFO] [./src/client/index.scss] 1.2 KiB {main} [built]
[INFO]     + 1522 hidden modules
[INFO] Child html-webpack-plugin for "index.html":
[INFO]      1 asset
[INFO]     Entrypoint undefined = index.html
[INFO]     [./node_modules/html-webpack-plugin/lib/loader.js!./public/index.html] 603 bytes {0} [built]
[INFO]     [./node_modules/webpack/buildin/global.js] (webpack)/buildin/global.js 472 bytes {0} [built]
[INFO]     [./node_modules/webpack/buildin/module.js] (webpack)/buildin/module.js 497 bytes {0} [built]
[INFO]         + 1 hidden module
[INFO]
[INFO] --- maven-resources-plugin:3.1.0:resources (default-resources) @ presentation-server ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/jonesn/IdeaProjects/egeria/open-metadata-implementation/user-interfaces/presentation-server/src/main/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ presentation-server ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-resources-plugin:3.1.0:testResources (default-testResources) @ presentation-server ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/jonesn/IdeaProjects/egeria/open-metadata-implementation/user-interfaces/presentation-server/src/test/resources
[INFO]
[INFO] --- jacoco-maven-plugin:0.8.5:prepare-agent (agent) @ presentation-server ---
[INFO] argLine set to -javaagent:/Users/jonesn/.m2/repository/org/jacoco/org.jacoco.agent/0.8.5/org.jacoco.agent-0.8.5-runtime.jar=destfile=/Users/jonesn/IdeaProjects/egeria/open-metadata-implementation/user-interfaces/presentation-server/target/jacoco.exec
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ presentation-server ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-surefire-plugin:3.0.0-M4:test (default-test) @ presentation-server ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-resources-plugin:3.1.0:copy-resources (copy-resources-from-node) @ presentation-server ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 3 resources
[INFO]
[INFO] --- maven-jar-plugin:3.2.0:jar (default-jar) @ presentation-server ---
[INFO] Building jar: /Users/jonesn/IdeaProjects/egeria/open-metadata-implementation/user-interfaces/presentation-server/target/presentation-server-2.0-SNAPSHOT.jar
[INFO]
[INFO] --- maven-source-plugin:3.2.1:jar-no-fork (attach-sources) @ presentation-server ---
[INFO] No sources in project. Archive not created.
[INFO]
[INFO] --- maven-javadoc-plugin:3.2.0:jar (attach-javadocs) @ presentation-server ---
[INFO]
[INFO] --- jacoco-maven-plugin:0.8.5:report (report) @ presentation-server ---
[INFO] Skipping JaCoCo execution due to missing execution data file.
[INFO]
[INFO] --- maven-dependency-plugin:3.1.2:analyze-only (analyze) @ presentation-server ---
[INFO] No dependency problems found
[INFO]
[INFO] --- maven-dependency-plugin:3.1.2:analyze-dep-mgt (depmgmt) @ presentation-server ---
[INFO] Found Resolved Dependency/DependencyManagement mismatches:
[INFO] 	None
[INFO]
[INFO] --- maven-install-plugin:3.0.0-M1:install (default-install) @ presentation-server ---
[INFO] Installing /Users/jonesn/IdeaProjects/egeria/open-metadata-implementation/user-interfaces/presentation-server/target/presentation-server-2.0-SNAPSHOT.jar to /Users/jonesn/.m2/repository/org/odpi/egeria/presentation-server/2.0-SNAPSHOT/presentation-server-2.0-SNAPSHOT.jar
[INFO] Installing /Users/jonesn/IdeaProjects/egeria/open-metadata-implementation/user-interfaces/presentation-server/pom.xml to /Users/jonesn/.m2/repository/org/odpi/egeria/presentation-server/2.0-SNAPSHOT/presentation-server-2.0-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:07 min
[INFO] Finished at: 2020-06-04T17:52:17+01:00
[INFO] ------------------------------------------------------------------------
```

**Comment by planetf1** [comment]
- also added package.json as I think this is needed on the server we deploy to, in order to run 'npm install' ?

**Comment by grahamwallis** [comment]
I am no expert I this area - but I have the following initial comments:
* The build shown above is in development mode.
* We should include the RAT check for the files that *should* have licenses.

**Comment by planetf1** [comment]
* Initial build will run webpack as @davidradl @grahamwallis are running
  - This may break if changes are made in webpack, since from maven we cannot directly run the commands in package.json due to platform and environment dependency

* The app is not added into the maven artifact for distribution
 - This is probably desirable in future -- I am aware some teams are taking maven artifacts to compose their runtime environment at a per-module level

* I've created a new assembly file for the files we need to place on the node server -- since this is running a different stack to the egeria server AND we are still working through what we really want, this keeps it simple/isolated -- we can add to the main egeria distro at a later date
 - This is highly geared to the current dev layout. Ideally more of the logic here should be in the node processing ie through webpack
 - the output is placed in `open-metadata-distribution/open-metadata-assemblies/target/egeria-2.0-SNAPSHOT-presentation-server/egeria-presentation-server-2.0-SNAPSHOT/`
 - You can run with `npm install && node src/server/index.js` - but this is somewhat subject to what is installed locally . May be suitable for UI devs but problematic for general use
* I added a docker image which will launch the UI using a standard alpine based node server image
 - this can be run currently with `$ docker run -p 8091:8091 planetf1/egeria-presentation-server:2.0-SNAPSHOT`
 - once in master `$ docker run -p 8091:8091 odpi/egeria-presentation-server:2.0-SNAPSHOT` should work
 - The UI can be viewed via https://localhost:8091/aaa/login

There's a lot to consider
 - dev vs production
 - driving webpack
 - RAT check (as per @grahamwallis comment above)
 - version updates
 - moving more logic out of the assembly into webpack
 - what's in the distribution(s)
 - what we call the docker image
 - adding docker images into the lab helm chart & linking up the configuration
 - correcting the certificates

However I propose that on the basis this is a first step, let's handle the followon updates in a new PR ? Assuming this works...

**Comment by grahamwallis** [comment]
Hi @planetf1 - I think it would be good to get the latest changes merged. Many thanks for your work on this.

File: open-metadata-implementation/user-interfaces/presentation-server/pom.xml [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                                    </includes>
+                                </resource>
+                            </resources>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin-->
+
+
+            <!-- TODO: Ensure anything not cleaned up by node is cleaned out here -->
```
Comment (davidradl) [inline open-metadata-implementation/user-interfaces/presentation-server/pom.xml:126]: Is this still a TODO?

File: open-metadata-resources/open-metadata-deployment/docker/presentation-server/pom.xml [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <scm>
+        <connection>scm:git:git://github.com/odpi/egeria.git</connection>
+        <developerConnection>scm:git:ssh://github.com/odpi/egeria.git</developerConnection>
+        <url>http://github.com/odpi/egeria/tree/master</url>
+    </scm>
+
+    <name>Docker image - Egeria</name>
```
Comment (davidradl) [inline open-metadata-resources/open-metadata-deployment/docker/presentation-server/pom.xml:23]: Shouldn't we call this something to do with presentation server

File: open-metadata-resources/open-metadata-deployment/docker/presentation-server/pom.xml [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    <description>
+        Samples, Helm charts to deploy Egeria and associated components such as docker image build scripts
```
Comment (davidradl) [inline open-metadata-resources/open-metadata-deployment/docker/presentation-server/pom.xml:25]: Shouldn't describe presentation server here

**Pull request merged**